mountpoint-s3/src/fs.rs

```diff
@@ -171,40 +185,69 @@ impl<Client: ObjectClient> UploadState<Client> {
                             error!(?err, ?key, "error updating the inode status");
                         }
                     }
-                    Self::Failed(_) | Self::Completed => unreachable!("checked above"),
+                    UploadState::NotStarted { .. } | Self::Failed(_) | Self::Completed => unreachable!("checked above"),
                 };
                 Err(e.into())
             }
         }
     }
 
-    async fn complete(&mut self, key: &str, ignore_if_empty: bool, pid: Option<u32>) -> Result<(), Error> {
-        match self {
-            Self::InProgress { request, handle } => {
-                if ignore_if_empty && request.size() == 0 {
-                    trace!(key, "not completing upload because file is empty");
-                    return Ok(());
-                }
-                if let Some(pid) = pid {
+    async fn complete(
+        &mut self,
+        key: &str,
+        ignore_if_empty: bool,
+        pid: Option<u32>,
+        uploader: &Uploader<Client>,
+    ) -> Result<(), Error> {
+        let (request_size, open_pid) = match self {
+            Self::NotStarted {
```

Comment:
The NotStarted case adds a lot of redundant code in all methods. It would be nice to refactor at least part of it. Or even better to avoid this case altogether, if possible.